### ISSUE TITLE
InstrumentationHelper.HasPdb always returns false for deterministic source paths

### ISSUE THREAD

**Issue by sharwell** [tonerdo/coverlet#363]
When `/p:DeterministicSourcePaths=true` is set, `InstrumentationHelper.HasPdb` looks for PDB files in the wrong location. These files are never found, so instrumentation is not performed.

https://github.com/tonerdo/coverlet/blob/062b907735901bdf3ad9d4d40953f53a4fe3f364/src/coverlet.core/Helpers/InstrumentationHelper.cs#L74

**Comment by ViceIce** [comment]
Additionally, it should have a `try ... catch` block, because `peReader.ReadDebugDirectory()` can fail.

Eg [Oracle.ManagedDataAccess 18.3.0](https://www.nuget.org/packages/Oracle.ManagedDataAccess/18.3.0) is failing.

**Comment by MarcoRossignoli** [comment]
> Additionally, it should have a try ... catch block, because peReader.ReadDebugDirectory() can fail.

Fixed in https://github.com/tonerdo/coverlet/pull/367/files

**Comment by MarcoRossignoli** [comment]
@sharwell is there a "simple" way to repro this(maybe some simple example on some project ie winform etc...)?

**Comment by sharwell** [comment]
@MarcoRossignoli If you have WinForms, you should be able to simply set an environment variable `DeterministicSourcePaths=true` and the build will pick it up.

**Comment by MarcoRossignoli** [comment]
.net 3.0?

**Comment by MarcoRossignoli** [comment]
@sharwell can you share a command to repro? I cloned windowsform and printed which modules is skipped
```
C:\git\winforms\src\System.Windows.Forms.Design\tests\UnitTests (master -> origin)
λ dotnet test /p:CollectCoverage=true
```
But with or without `set DeterministicSourcePaths=true` I get same list

**Comment by sharwell** [comment]
@MarcoRossignoli Here's a repro case:

1. Check out dotnet/roslyn-analyzers#3028
2. Build with `.\Build.cmd -rebuild /p:DeterministicSourcePaths=true`
3. Test with `.\Test.cmd /p:Coverage=True`

Expected results:

Coverage files in **artifacts\\coverage** have coverage information.

Actual results:

Coverage files are empty.

**Comment by MarcoRossignoli** [comment]
great thank's I'll take a look asap some busy weeks!

**Comment by MarcoRossignoli** [comment]
Hi @tmat I'm working on this issue and found https://github.com/dotnet/sourcelink/issues/91#issuecomment-400787926

Btw I don't undestand why it doesn't work on current repo, I've build with
```
D:\git\coverlet\test\coverlet.core.tests (master -> origin)
λ dotnet build ..\..\ /p:DeterministicSourcePaths=true
```
When I debug and try read debug directory for a referenced lib(from test) there is an "incomplete" path for source file inside pdb
![image](https://user-images.githubusercontent.com/7894084/69481738-0e48ac80-0e14-11ea-812b-95a23668cef6.png)

`coverlet.tests.projectsample.empty.dll` is referenced by test project and imports `Microsoft.SourceLink.GitHub` package from https://github.com/MarcoRossignoli/coverlet/blob/issue_363/Directory.Build.props#L16

If I understood well `Microsoft.SourceLink.GitHub` should load `SourceRoot` during build with correct path.

@sharwell the problem is here...if compiled with `/p:DeterministicSourcePaths=true` the paths reported inside pdbs are all relative to `/ _/...` and we fail to find local source.

**Comment by sharwell** [comment]
@MarcoRossignoli yes, the feature improves build determinism by hiding the path `D:\git\coverlet\` that can vary across build machines.

**Comment by MarcoRossignoli** [comment]
Is there a standard "start with" prefix in case of deterministic build?I mean can we check `/ _/`?

**Comment by sharwell** [comment]
Yes, the prefix is `/_/`.

**Comment by MarcoRossignoli** [comment]
ok thank's Sam!

**Comment by tmat** [comment]
It's not always `/_/` though. Each SourceRoot has a unique prefix `/_/`, `/_1/`, `/_2/`, etc. For example, when source files are linked from source packages each source package dir gets a distinct SourceRoot.

**Comment by tmat** [comment]
In any case, you should not rely on a certain prefix or prefixes. Instead, read the `$(PathMap)` msbuild property that's being passed to the compiler (Csc task). This map defines mapping that the compiler applies to all paths used in the compilation (source paths, pdb path, etc.).

This is an example of a `PathMap` value for `dotnet/symreader-converter` repo, which is using source packages:

```
F:\workspace\_work\1\s\=/_/,F:\workspace\_work\1\s\.packages\microsoft.codeanalysis.pooledobjects\3.3.0-beta3-19407-01\contentFiles\cs\netstandard1.3\=/_1/,F:\workspace\_work\1\s\.packages\microsoft.codeanalysis.debugging\3.3.0-beta3-19407-01\contentFiles\cs\netstandard1.3\=/_2/
```

**Comment by sharwell** [comment]
@tmat Interesting. I never saw any path but `/_/`. It should be easy enough to use the path map though 👍

**Comment by MarcoRossignoli** [comment]
We don't have always msbuild context available, coverlet could be used through vstest collectors(injected by test runtime)  and .net tool(console app that manually instrument dll after build and restore) where build is done, maybe change the path a bit to include possible 1,2 etc...I mean we should do something based only on produced artifact (pdb/dll)

**Comment by tmat** [comment]
@MarcoRossignoli That's unfortunate. Other custom build targets can specify arbitrary path mapping. 

Let's see if we can come up with a better solution than hard-coding some patterns. At what point do you actually need to read the source files? I'd assume you don't need to do so for the instrumentation itself. So you can produce instrumented binaries using just the pdb and the dll as inputs. Then later when you report which lines are covered you need to read the source files. Correct?

**Comment by MarcoRossignoli** [comment]
@tmat we use source in more than one phase, I'll try to explain workflow:

Coverlet has got mainly two phase:

1) prepare modules where we instrument asm and persist a structure where we map an "index"(simple integer) to sequence points/branches candidates https://github.com/tonerdo/coverlet/blob/master/src/coverlet.core/Instrumentation/Instrumenter.cs#L440-L498

We have two "hit candidates" sequence points for lines and branch point for branches.

In this phase we use Cecil to read and instrument asm, we use also pdb(DebugInformation) to get documents name and prepare "the structure" to persist and reload after testing.
In addition we need to know if a "pdb+local source file" exists to check if an asm is "instrumentable". 
For instance when we read all asm from target folder we need to skip xunit dll or other dependeces for what we don't have source+pdb(it's not our code to cover), to do that we check if dll has got a pdb on local disk or an embedded one and if it's so we check that the "source file name" inside pdb are present on disk(for instance xunit has got an embedded pdb but we won't find source files on local disk and we'll skip without any need to use "filters", one of the most requested feature by users).

https://github.com/tonerdo/coverlet/blob/4f3f25fda9badcb50aac83e182c79bbc4c7ff8df/src/coverlet.core/Instrumentation/Instrumenter.cs#L69+L110

2) at the end of the test a file with hits will persisted and we load this file and togheter with persisted structure(modules/documents/hitscandidates mappings) we "account" hits 

https://github.com/tonerdo/coverlet/blob/master/src/coverlet.core/Coverage.cs#L352-L385

We'll use file name inside persisted structure also to populate reports file(cobertura etc...) so tools like reportgenerator can show coverage on sources.

So to sum up we use "source file name path" everywhere.

Feel free to ask if it's not clear.

**Comment by MarcoRossignoli** [comment]
@tmat I'm trying with PathMap property on msbuild for now, some observation.

PathMap are generated and passed during "build" but "test" run after build so I don't get any PathMap data. BTW if I pass `/p:DeterministicSourcePaths=true` with `dotnet test /p:CollectCoverage=true /p:DeterministicSourcePaths=true` I get some map data, now does it make sense?Or map paths could be different from build one?

**Comment by olivier-spinelli** [comment]
@tmat Did you (Roslyn team?) consider systematically embedding the $(PathMap) as as either a standardized assembly attribute or as a resource (with a special name)?

**Comment by clairernovotny** [comment]
Any update on this one? We're trying to push more assemblies to be deterministic and it's breaking coverlet

**Comment by clairernovotny** [comment]
What if we check to see if the pdb is alongside the DLL and not rely on the path from the pdb file?

**Comment by clairernovotny** [comment]
Proposal:

1. Short term: Only coverlet.collector or coverlet.msbuild will support DeterministicSourcePaths. It can do so by adding a target that writes the PathMap's to the output directory alongside the target dll's. This target may need to be added at the Directory Props level to work, next to the Source Link reference.
2. Better: SourceLink / the SDK writes this out built-in.

In both cases, build tools that need to read this data can open the PathMap file and do the translation needed.

Thoughts?

**Comment by clairernovotny** [comment]
Bad workaround: Split the build/test into two passes.

1. Build/Pack whatever is needed using deterministic settings
2. In the test invocation run `/t:rebuild /p:ContinuousIntegrationBuild=false` to force a binary rebuild of the test project and its dependencies without the source paths being messed up.

**Comment by MarcoRossignoli** [comment]
@clairernovotny thank's a lot for the ideas, your first short term idea was one in my mind, I was waiting @tmat opinion on it to understand if there are better choice to avoid to waste time on "suboptimal" implementation.

> Better: SourceLink / the SDK writes this out built-in.

This should be done at the end.

**Comment by MarcoRossignoli** [comment]
Some notes 
https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets
https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/MSBuildTask/MapSourceRoots.cs

**Comment by clairernovotny** [comment]
I’ve put together a sample showing how to get the path mappings per referenced project:

https://github.com/clairernovotny/DeterministicBuilds/tree/retrieve-pathmap

Here’s what to look for: 
-	https://github.com/clairernovotny/DeterministicBuilds/blob/retrieve-pathmap/Directory.Build.targets#L12-L22
o	Workaround target that needs to be in a Directory.Build.targets file. This returns the Source Roots.
-	https://github.com/clairernovotny/DeterministicBuilds/blob/retrieve-pathmap/SomeLibrary.Tests/SomeLibrary.Tests.csproj#L12-L26
o	Would be supplied in the coverlet targets that get included in the test project. That gets the source maps for each reference and then transforms them to be by project.


I output this with  <Message Text="@(_byProject->'%(Identity) :: %(OriginalPath)=%(MappedPath)')" Importance="high" /> to show how to unpack it. By project, I’m outputting each original path and its mapped path. Each project may have more than one path mapping…

You can run this locally with dotnet build /p:TF_BUILD=true on the test project. The test project references a class library, which references another, showing transitive project references work.

On my machine, it writes the following message:

  D:\dev\ClassLibrary1\ClassLibrary1\ClassLibrary1.csproj :: D:\dev\ClassLibrary1\=/_/;D:\dev\ClassLibrary1\ClassLibrar
y2\ClassLibrary2.csproj :: D:\dev\ClassLibrary1\=/_/

**Comment by MarcoRossignoli** [comment]
@clairernovotny something strange on test repo seem that some files are not in git index 
```bash
C:\git\DeterministicBuilds (master -> origin)
λ ls
ClassLibrary1/  ClassLibrary1.sln  ClassLibrary1.sln.DotSettings.user  Directory.Build.props  Directory.Build.targets  LICENSE  README.md
```
I downloaded zip version, I get error 
```
C:\git\DeterministicBuilds-retrieve-pathmap
λ dotnet build /p:TF_BUILD=true
Microsoft (R) Build Engine version 16.6.0-preview-20112-08+77da97f69 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 203,39 ms for C:\git\DeterministicBuilds-retrieve-pathmap\ClassLibrary2\ClassLibrary2.csproj.
  Restore completed in 203,37 ms for C:\git\DeterministicBuilds-retrieve-pathmap\ClassLibrary1\ClassLibrary1.csproj.
  Restore completed in 203,39 ms for C:\git\DeterministicBuilds-retrieve-pathmap\SomeLibrary.Tests\SomeLibrary.Tests.csproj.
  You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview
C:\Users\Marco\.nuget\packages\microsoft.build.tasks.git\1.0.0\build\Microsoft.Build.Tasks.Git.targets(24,5): warning : Unable to locate repository with working directory that contains directory 'C:\git\DeterministicBuilds-retrieve-pathmap\ClassLibrary2'. [C:\git\DeterministicBui
lds-retrieve-pathmap\ClassLibrary2\ClassLibrary2.csproj]
C:\Program Files\dotnet\sdk\5.0.100-preview.1.20155.7\Roslyn\Microsoft.Managed.Core.targets(104,5): error : SourceRoot items must include at least one top-level (not nested) item when DeterministicSourcePaths is true [C:\git\DeterministicBuilds-retrieve-pathmap\ClassLibrary2\Clas
sLibrary2.csproj]

Build FAILED.

C:\Users\Marco\.nuget\packages\microsoft.build.tasks.git\1.0.0\build\Microsoft.Build.Tasks.Git.targets(24,5): warning : Unable to locate repository with working directory that contains directory 'C:\git\DeterministicBuilds-retrieve-pathmap\ClassLibrary2'. [C:\git\DeterministicBui
lds-retrieve-pathmap\ClassLibrary2\ClassLibrary2.csproj]
C:\Program Files\dotnet\sdk\5.0.100-preview.1.20155.7\Roslyn\Microsoft.Managed.Core.targets(104,5): error : SourceRoot items must include at least one top-level (not nested) item when DeterministicSourcePaths is true [C:\git\DeterministicBuilds-retrieve-pathmap\ClassLibrary2\Clas
sLibrary2.csproj]
    1 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.91
```
What am I missing?

**Comment by clairernovotny** [comment]
Weird...I just did a fresh clone and checkout the branch and it worked for me?

**Comment by MarcoRossignoli** [comment]
> did you check out the retrieve-pathmap

Thanks solved!

**Comment by MarcoRossignoli** [comment]
Plan
- [ ] msbuild support(wip https://github.com/tonerdo/coverlet/compare/master...MarcoRossignoli:deterministicbuild)
- [ ] collector support
- [ ] add new integration tests
- [ ] add workaround documentation

**Comment by clairernovotny** [comment]
The latest nightly I see for coverlet.collectors was published in March 30, I don't see one from last night?

**Comment by MarcoRossignoli** [comment]
Just pushed there was an issue with pushing and also after `2.8.1` release the order now is not preserved, btw this is the package https://www.myget.org/feed/coverlet-dev/package/nuget/coverlet.msbuild/2.8.1-g4e177f0c57

**Comment by clairernovotny** [comment]
Maybe we should talk versioning.... 

For prerelease builds, I tend to use a preview tag.

In the version.json, you can use `2.8.2-preview.{height}` and that means that every build gets a new version. When you're ready to release, you can change it to `2.8.2`, do the release, then change it to `2.8.3-preview.{height}` again.

Another thing to consider is unifying the versions in the repo to make it easier. Take the highest version, move a single version.json to the root and then bump a major. If you do Major.Minor, then each build gets it's own patch version automatically. You could also do Major.Minor.0-preview.{height} as well.

**Comment by clairernovotny** [comment]
Also, I'm only using `coverlet.collector` and I only see a single version there: 
![image](https://user-images.githubusercontent.com/1427284/78925162-f98a4f80-7a68-11ea-998a-906ad35878b2.png)

**Comment by MarcoRossignoli** [comment]
> In the version.json, you can use 2.8.2-preview.{height}

Yes this is my idea, since nightly we didn't have minor update and it worked untile now, so that was an error, thanks.

> Another thing to consider is unifying the versions in the repo to make it easier

I'm not sure about it...at the moment we have 3 drivers(.net tool,msbuild,collectors) and not all version supports all features, have single version could confuse users.

**Comment by clairernovotny** [comment]
> I'm not sure about it...at the moment we have 3 drivers(.net tool,msbuild,collectors) and not all version supports all features, have single version could confuse users.

FWIW, I think it could be okay to keep drivers at the same version long as what's in each is documented somewhere. If they all tend to rely on an underlying core library, then that could be what drives the main version. Totally up to you though of course.

**Comment by MarcoRossignoli** [comment]
> FWIW, I think it could be okay to keep drivers at the same version long as what's in each is documented somewhere. If they all tend to rely on an underlying core library, then that could be what drives the main version

Thank's for the advice, I'll think about cc: @tonerdo

**Comment by MarcoRossignoli** [comment]
@clairernovotny nightly should be ok I tool a look at your Rx.NET repo and I prefer leave on version also commit id so users can do fast check without nbgv tool.

**Comment by MarcoRossignoli** [comment]
@clairernovotny @tmat support/tests are ok for now, does it make sense work on documentation related to https://github.com/reactiveui/refit/pull/902?

**Comment by clairernovotny** [comment]
@MarcoRossignoli What documentation for that issue? Not quite following but all docs are generally useful!

**Comment by MarcoRossignoli** [comment]
I mean the workaround to do to support deterministic build coverage, but it's not clear to me if we can release or if it doesn't work, related to your issue on reactive repo.

**Comment by clairernovotny** [comment]
I need @tmat to take a closer look at what's going on the the VSTest case.
